{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DeNovo Negative Dataset Sampling for HP-PPI\n",
    "\n",
    "Dissimilarity-based negative sampling for non-interacting protein pairs\n",
    "\n",
    "Reference:\n",
    "\n",
    ">Eid, F. E., Elhefnawi, M., & Heath, L. S. (2016). DeNovo: Virus-host sequence-based protein-protein interaction prediction. *Bioinformatics*, **32**(8), 1144â€“1150. https://doi.org/10.1093/bioinformatics/btv737"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import random\n",
    "\n",
    "from time import time\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "from multiprocessing import Pool\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import PairwiseAligner\n",
    "from Bio.SubsMat.MatrixInfo import blosum30\n",
    "\n",
    "# Print status of datasets\n",
    "def print_status(df):\n",
    "    \n",
    "    # For each pathogen\n",
    "    for pathogen in sorted(set(df.Pathogen)):\n",
    "        df_patho = df[df.Pathogen == pathogen]\n",
    "        i = len(df_patho)\n",
    "        p = len(set(df_patho.Pathogen_Uniprot_ID))\n",
    "        h = len(set(df_patho.Human_Uniprot_ID))\n",
    "        print('%s:\\n%i non-interacting pairs involving %i pathogen proteins and %i human proteins\\n' % (pathogen, i, p, h))\n",
    "    \n",
    "    # Total\n",
    "    i = len(df)\n",
    "    p = len(set(df.Pathogen_Uniprot_ID))\n",
    "    h = len(set(df.Human_Uniprot_ID))\n",
    "    print('TOTAL:\\n%i non-interacting pairs involving %i pathogen proteins and %i human proteins\\n' % (i, p, h))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = os.path.dirname(os.getcwd())\n",
    "dir_in = dir_out = os.path.join(parent_dir, 'data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up the sequences of pathogen proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtained 2284 pathogen proteins for alignment\n"
     ]
    }
   ],
   "source": [
    "# Read positive dataset\n",
    "f_in = os.path.join(dir_in, 'positive_pairs.tsv')\n",
    "df_pos = pd.read_csv(f_in, sep='\\t')\n",
    "\n",
    "# Get pathogen protein IDs for alignment\n",
    "prot_IDs = list(set(df_pos.Pathogen_Uniprot_ID))\n",
    "prot_IDs.sort(reverse=True)\n",
    "\n",
    "print('Obtained %i pathogen proteins for alignment' % len(prot_IDs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtained 2284 pathogen protein sequences\n"
     ]
    }
   ],
   "source": [
    "# Extract pathogen protein sequences\n",
    "f_in = os.path.join(dir_in, 'uniprot_sequences.fasta')\n",
    "seqdict = {} # store sequences in a dict\n",
    "\n",
    "for record in SeqIO.parse(f_in, 'fasta'):\n",
    "    prot_id = record.id.split('|')[1]\n",
    "    \n",
    "    # Check if record is in the pathogen IDs\n",
    "    if prot_id in prot_IDs:\n",
    "        seqdict[prot_id] = record.seq\n",
    "\n",
    "print('Obtained %i pathogen protein sequences' % len(seqdict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All-vs-all global alignment of pathogen proteins\n",
    "\n",
    "May skip to the next section (`Negative Sampling`) if alignment scores file already exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment algorithm: Needleman-Wunsch\n"
     ]
    }
   ],
   "source": [
    "# Set up aligner\n",
    "aligner = PairwiseAligner()\n",
    "\n",
    "# Gap scoring: linear, 8 (as in MATLAB, which is the platform used by Eid et al., 2016)\n",
    "# Matrix: BLOSUM30 to capture distant relationships\n",
    "\n",
    "aligner.gap_score = -10\n",
    "aligner.substitution_matrix = blosum30\n",
    "print('Alignment algorithm: %s' % aligner.algorithm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated 2609470 pairs for alignment\n"
     ]
    }
   ],
   "source": [
    "# Generate pairs of proteins to be aligned\n",
    "pairs = [pair for pair in itertools.combinations_with_replacement(prot_IDs, 2)]\n",
    "print('Generated %i pairs for alignment' % len(pairs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2284, 2284)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <th>Q9ZC81</th>\n",
       "      <th>Q9ZC68</th>\n",
       "      <th>Q9ZC63</th>\n",
       "      <th>Q9ZC61</th>\n",
       "      <th>Q9ZC54</th>\n",
       "      <th>Q9ZC51</th>\n",
       "      <th>Q9ZC50</th>\n",
       "      <th>Q9ZC30</th>\n",
       "      <th>Q9ZC29</th>\n",
       "      <th>...</th>\n",
       "      <th>A0A0F7R517</th>\n",
       "      <th>A0A0F7R4Q5</th>\n",
       "      <th>A0A0F7R444</th>\n",
       "      <th>A0A0F7R416</th>\n",
       "      <th>A0A0F7R3Y0</th>\n",
       "      <th>A0A0F7R3S6</th>\n",
       "      <th>A0A0F7R3M8</th>\n",
       "      <th>A0A0F7R3M4</th>\n",
       "      <th>A0A0F7R3D8</th>\n",
       "      <th>A0A0F7R2X5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC81</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC68</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC63</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC61</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Q9ZFR9  Q9ZC81  Q9ZC68  Q9ZC63  Q9ZC61  Q9ZC54  Q9ZC51  Q9ZC50  \\\n",
       "Q9ZFR9     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC81     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC68     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC63     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC61     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "        Q9ZC30  Q9ZC29     ...      A0A0F7R517  A0A0F7R4Q5  A0A0F7R444  \\\n",
       "Q9ZFR9     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC81     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC68     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC63     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC61     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "\n",
       "        A0A0F7R416  A0A0F7R3Y0  A0A0F7R3S6  A0A0F7R3M8  A0A0F7R3M4  \\\n",
       "Q9ZFR9         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC81         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC68         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC63         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC61         0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "        A0A0F7R3D8  A0A0F7R2X5  \n",
       "Q9ZFR9         0.0         0.0  \n",
       "Q9ZC81         0.0         0.0  \n",
       "Q9ZC68         0.0         0.0  \n",
       "Q9ZC63         0.0         0.0  \n",
       "Q9ZC61         0.0         0.0  \n",
       "\n",
       "[5 rows x 2284 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize zeros DataFrame for storing alignment scores\n",
    "scores = pd.DataFrame(0., index=prot_IDs, columns=prot_IDs)\n",
    "\n",
    "scores.shape\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get the alignment score\n",
    "def get_alignment_score(pair):\n",
    "    score = aligner.score(*map(seqdict.get, pair))\n",
    "    \n",
    "    # Return both proteins along with the alignment score\n",
    "    return (*pair, score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment duration: 13.6411 minutes\n"
     ]
    }
   ],
   "source": [
    "# Perform alignments with parallelization\n",
    "with Pool(6) as p:\n",
    "    t0 = time()\n",
    "    \n",
    "    results = p.map(get_alignment_score, pairs)\n",
    "    \n",
    "    t = time() - t0\n",
    "    print('Alignment duration: %.4f minutes' % (t/60))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2609470/2609470 [00:58<00:00, 44316.21it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <th>Q9ZC81</th>\n",
       "      <th>Q9ZC68</th>\n",
       "      <th>Q9ZC63</th>\n",
       "      <th>Q9ZC61</th>\n",
       "      <th>Q9ZC54</th>\n",
       "      <th>Q9ZC51</th>\n",
       "      <th>Q9ZC50</th>\n",
       "      <th>Q9ZC30</th>\n",
       "      <th>Q9ZC29</th>\n",
       "      <th>...</th>\n",
       "      <th>A0A0F7R517</th>\n",
       "      <th>A0A0F7R4Q5</th>\n",
       "      <th>A0A0F7R444</th>\n",
       "      <th>A0A0F7R416</th>\n",
       "      <th>A0A0F7R3Y0</th>\n",
       "      <th>A0A0F7R3S6</th>\n",
       "      <th>A0A0F7R3M8</th>\n",
       "      <th>A0A0F7R3M4</th>\n",
       "      <th>A0A0F7R3D8</th>\n",
       "      <th>A0A0F7R2X5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <td>3953.0</td>\n",
       "      <td>-1432.0</td>\n",
       "      <td>-126.0</td>\n",
       "      <td>-1205.0</td>\n",
       "      <td>-1818.0</td>\n",
       "      <td>-659.0</td>\n",
       "      <td>-150.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>-1549.0</td>\n",
       "      <td>-451.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1063.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>-1447.0</td>\n",
       "      <td>-892.0</td>\n",
       "      <td>-927.0</td>\n",
       "      <td>-1989.0</td>\n",
       "      <td>-684.0</td>\n",
       "      <td>-25.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>-1475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC81</th>\n",
       "      <td>-1432.0</td>\n",
       "      <td>2368.0</td>\n",
       "      <td>-914.0</td>\n",
       "      <td>-74.0</td>\n",
       "      <td>-48.0</td>\n",
       "      <td>-353.0</td>\n",
       "      <td>-1996.0</td>\n",
       "      <td>-1305.0</td>\n",
       "      <td>-3991.0</td>\n",
       "      <td>-497.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-74.0</td>\n",
       "      <td>-1452.0</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-222.0</td>\n",
       "      <td>-111.0</td>\n",
       "      <td>-197.0</td>\n",
       "      <td>-419.0</td>\n",
       "      <td>-971.0</td>\n",
       "      <td>-1433.0</td>\n",
       "      <td>-6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC68</th>\n",
       "      <td>-126.0</td>\n",
       "      <td>-914.0</td>\n",
       "      <td>3479.0</td>\n",
       "      <td>-705.0</td>\n",
       "      <td>-1279.0</td>\n",
       "      <td>-231.0</td>\n",
       "      <td>-504.0</td>\n",
       "      <td>-33.0</td>\n",
       "      <td>-2160.0</td>\n",
       "      <td>-100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-530.0</td>\n",
       "      <td>-152.0</td>\n",
       "      <td>-939.0</td>\n",
       "      <td>-455.0</td>\n",
       "      <td>-512.0</td>\n",
       "      <td>-1481.0</td>\n",
       "      <td>-262.0</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>-90.0</td>\n",
       "      <td>-977.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC63</th>\n",
       "      <td>-1205.0</td>\n",
       "      <td>-74.0</td>\n",
       "      <td>-705.0</td>\n",
       "      <td>2311.0</td>\n",
       "      <td>-328.0</td>\n",
       "      <td>-186.0</td>\n",
       "      <td>-1789.0</td>\n",
       "      <td>-978.0</td>\n",
       "      <td>-3702.0</td>\n",
       "      <td>-307.0</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>-1210.0</td>\n",
       "      <td>-62.0</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-417.0</td>\n",
       "      <td>-216.0</td>\n",
       "      <td>-816.0</td>\n",
       "      <td>-1139.0</td>\n",
       "      <td>-18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC61</th>\n",
       "      <td>-1818.0</td>\n",
       "      <td>-48.0</td>\n",
       "      <td>-1279.0</td>\n",
       "      <td>-328.0</td>\n",
       "      <td>2246.0</td>\n",
       "      <td>-702.0</td>\n",
       "      <td>-2452.0</td>\n",
       "      <td>-1679.0</td>\n",
       "      <td>-4422.0</td>\n",
       "      <td>-840.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-293.0</td>\n",
       "      <td>-1841.0</td>\n",
       "      <td>-140.0</td>\n",
       "      <td>-486.0</td>\n",
       "      <td>-414.0</td>\n",
       "      <td>-32.0</td>\n",
       "      <td>-696.0</td>\n",
       "      <td>-1390.0</td>\n",
       "      <td>-1741.0</td>\n",
       "      <td>-108.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Q9ZFR9  Q9ZC81  Q9ZC68  Q9ZC63  Q9ZC61  Q9ZC54  Q9ZC51  Q9ZC50  \\\n",
       "Q9ZFR9  3953.0 -1432.0  -126.0 -1205.0 -1818.0  -659.0  -150.0    57.0   \n",
       "Q9ZC81 -1432.0  2368.0  -914.0   -74.0   -48.0  -353.0 -1996.0 -1305.0   \n",
       "Q9ZC68  -126.0  -914.0  3479.0  -705.0 -1279.0  -231.0  -504.0   -33.0   \n",
       "Q9ZC63 -1205.0   -74.0  -705.0  2311.0  -328.0  -186.0 -1789.0  -978.0   \n",
       "Q9ZC61 -1818.0   -48.0 -1279.0  -328.0  2246.0  -702.0 -2452.0 -1679.0   \n",
       "\n",
       "        Q9ZC30  Q9ZC29     ...      A0A0F7R517  A0A0F7R4Q5  A0A0F7R444  \\\n",
       "Q9ZFR9 -1549.0  -451.0     ...         -1063.0        52.0     -1447.0   \n",
       "Q9ZC81 -3991.0  -497.0     ...           -74.0     -1452.0        -7.0   \n",
       "Q9ZC68 -2160.0  -100.0     ...          -530.0      -152.0      -939.0   \n",
       "Q9ZC63 -3702.0  -307.0     ...            25.0     -1210.0       -62.0   \n",
       "Q9ZC61 -4422.0  -840.0     ...          -293.0     -1841.0      -140.0   \n",
       "\n",
       "        A0A0F7R416  A0A0F7R3Y0  A0A0F7R3S6  A0A0F7R3M8  A0A0F7R3M4  \\\n",
       "Q9ZFR9      -892.0      -927.0     -1989.0      -684.0       -25.0   \n",
       "Q9ZC81      -222.0      -111.0      -197.0      -419.0      -971.0   \n",
       "Q9ZC68      -455.0      -512.0     -1481.0      -262.0       -14.0   \n",
       "Q9ZC63        -5.0        14.0      -417.0      -216.0      -816.0   \n",
       "Q9ZC61      -486.0      -414.0       -32.0      -696.0     -1390.0   \n",
       "\n",
       "        A0A0F7R3D8  A0A0F7R2X5  \n",
       "Q9ZFR9        83.0     -1475.0  \n",
       "Q9ZC81     -1433.0        -6.0  \n",
       "Q9ZC68       -90.0      -977.0  \n",
       "Q9ZC63     -1139.0       -18.0  \n",
       "Q9ZC61     -1741.0      -108.0  \n",
       "\n",
       "[5 rows x 2284 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store alignment scores\n",
    "for prot1, prot2, score in tqdm(results):\n",
    "    scores.at[prot1, prot2] = score\n",
    "    scores.at[prot2, prot1] = score\n",
    "\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the scores DataFrame\n",
    "f_out = os.path.join(dir_out, 'blosum30_global_scores.tsv')\n",
    "scores.to_csv(f_out, sep='\\t', index=prot_IDs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Negative sampling\n",
    "\n",
    "Select non-interacting pairs based on dissimilarity between pathogen protein interactors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the scores DataFrame\n",
    "f_in = os.path.join(dir_in, 'blosum30_global_scores.tsv')\n",
    "scores = pd.read_csv(f_in, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removed outlier protein: Q81SN0\n",
      "Removed outlier protein: Q8CZU2\n",
      "Removed outlier protein: Q7CGD9\n",
      "No more outliers found.\n"
     ]
    }
   ],
   "source": [
    "# Designate outlier scores as NaN (protein with very low alignment score compared to others)\n",
    "# Keep removing outliers as long as it has minimum scores across x% of total pathogen proteins\n",
    "x = 0.99\n",
    "\n",
    "while True:\n",
    "    row_mins = scores.idxmin(axis=1)\n",
    "    outlier_id = row_mins.mode()[0] # a bug: axis=0 actually considers columns instead of rows\n",
    "    \n",
    "    n_outlier = Counter(row_mins)[outlier_id]\n",
    "    if n_outlier > int(x * len(prot_IDs)):\n",
    "        \n",
    "        # Remove all entries in the outlier column except for its own row\n",
    "        scores.loc[~(scores.index == outlier_id), outlier_id] = np.nan\n",
    "        print('Removed outlier protein: %s' % outlier_id)\n",
    "    \n",
    "    else:\n",
    "        print('No more outliers found.')\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <th>Q9ZC81</th>\n",
       "      <th>Q9ZC68</th>\n",
       "      <th>Q9ZC63</th>\n",
       "      <th>Q9ZC61</th>\n",
       "      <th>Q9ZC54</th>\n",
       "      <th>Q9ZC51</th>\n",
       "      <th>Q9ZC50</th>\n",
       "      <th>Q9ZC30</th>\n",
       "      <th>Q9ZC29</th>\n",
       "      <th>...</th>\n",
       "      <th>A0A0F7R517</th>\n",
       "      <th>A0A0F7R4Q5</th>\n",
       "      <th>A0A0F7R444</th>\n",
       "      <th>A0A0F7R416</th>\n",
       "      <th>A0A0F7R3Y0</th>\n",
       "      <th>A0A0F7R3S6</th>\n",
       "      <th>A0A0F7R3M8</th>\n",
       "      <th>A0A0F7R3M4</th>\n",
       "      <th>A0A0F7R3D8</th>\n",
       "      <th>A0A0F7R2X5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133568</td>\n",
       "      <td>0.132957</td>\n",
       "      <td>0.124964</td>\n",
       "      <td>0.141194</td>\n",
       "      <td>0.126610</td>\n",
       "      <td>0.168132</td>\n",
       "      <td>0.127884</td>\n",
       "      <td>0.296699</td>\n",
       "      <td>0.132290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.136215</td>\n",
       "      <td>0.125944</td>\n",
       "      <td>0.130057</td>\n",
       "      <td>0.122946</td>\n",
       "      <td>0.127637</td>\n",
       "      <td>0.132757</td>\n",
       "      <td>0.120328</td>\n",
       "      <td>0.130324</td>\n",
       "      <td>0.136757</td>\n",
       "      <td>0.129596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC81</th>\n",
       "      <td>0.201814</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.162020</td>\n",
       "      <td>0.084767</td>\n",
       "      <td>0.079700</td>\n",
       "      <td>0.115540</td>\n",
       "      <td>0.238703</td>\n",
       "      <td>0.178903</td>\n",
       "      <td>0.395115</td>\n",
       "      <td>0.133953</td>\n",
       "      <td>...</td>\n",
       "      <td>0.101059</td>\n",
       "      <td>0.182436</td>\n",
       "      <td>0.079440</td>\n",
       "      <td>0.098909</td>\n",
       "      <td>0.098512</td>\n",
       "      <td>0.070771</td>\n",
       "      <td>0.110741</td>\n",
       "      <td>0.165358</td>\n",
       "      <td>0.193495</td>\n",
       "      <td>0.077961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC68</th>\n",
       "      <td>0.152869</td>\n",
       "      <td>0.115360</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.107194</td>\n",
       "      <td>0.122468</td>\n",
       "      <td>0.111127</td>\n",
       "      <td>0.181665</td>\n",
       "      <td>0.131256</td>\n",
       "      <td>0.321323</td>\n",
       "      <td>0.119599</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117269</td>\n",
       "      <td>0.133606</td>\n",
       "      <td>0.112201</td>\n",
       "      <td>0.107268</td>\n",
       "      <td>0.112824</td>\n",
       "      <td>0.115185</td>\n",
       "      <td>0.105061</td>\n",
       "      <td>0.129916</td>\n",
       "      <td>0.143231</td>\n",
       "      <td>0.112091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC63</th>\n",
       "      <td>0.193307</td>\n",
       "      <td>0.085835</td>\n",
       "      <td>0.154311</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.089428</td>\n",
       "      <td>0.109499</td>\n",
       "      <td>0.230790</td>\n",
       "      <td>0.166654</td>\n",
       "      <td>0.383468</td>\n",
       "      <td>0.127083</td>\n",
       "      <td>...</td>\n",
       "      <td>0.097540</td>\n",
       "      <td>0.173346</td>\n",
       "      <td>0.081374</td>\n",
       "      <td>0.091124</td>\n",
       "      <td>0.094050</td>\n",
       "      <td>0.078381</td>\n",
       "      <td>0.103397</td>\n",
       "      <td>0.159618</td>\n",
       "      <td>0.182492</td>\n",
       "      <td>0.078383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC61</th>\n",
       "      <td>0.216280</td>\n",
       "      <td>0.084921</td>\n",
       "      <td>0.175481</td>\n",
       "      <td>0.093794</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.128165</td>\n",
       "      <td>0.256136</td>\n",
       "      <td>0.192913</td>\n",
       "      <td>0.412485</td>\n",
       "      <td>0.146354</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108844</td>\n",
       "      <td>0.197048</td>\n",
       "      <td>0.084115</td>\n",
       "      <td>0.108381</td>\n",
       "      <td>0.109326</td>\n",
       "      <td>0.065064</td>\n",
       "      <td>0.120763</td>\n",
       "      <td>0.180875</td>\n",
       "      <td>0.205023</td>\n",
       "      <td>0.081547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Q9ZFR9    Q9ZC81    Q9ZC68    Q9ZC63    Q9ZC61    Q9ZC54    Q9ZC51  \\\n",
       "Q9ZFR9  0.000000  0.133568  0.132957  0.124964  0.141194  0.126610  0.168132   \n",
       "Q9ZC81  0.201814  0.000000  0.162020  0.084767  0.079700  0.115540  0.238703   \n",
       "Q9ZC68  0.152869  0.115360  0.000000  0.107194  0.122468  0.111127  0.181665   \n",
       "Q9ZC63  0.193307  0.085835  0.154311  0.000000  0.089428  0.109499  0.230790   \n",
       "Q9ZC61  0.216280  0.084921  0.175481  0.093794  0.000000  0.128165  0.256136   \n",
       "\n",
       "          Q9ZC50    Q9ZC30    Q9ZC29     ...      A0A0F7R517  A0A0F7R4Q5  \\\n",
       "Q9ZFR9  0.127884  0.296699  0.132290     ...        0.136215    0.125944   \n",
       "Q9ZC81  0.178903  0.395115  0.133953     ...        0.101059    0.182436   \n",
       "Q9ZC68  0.131256  0.321323  0.119599     ...        0.117269    0.133606   \n",
       "Q9ZC63  0.166654  0.383468  0.127083     ...        0.097540    0.173346   \n",
       "Q9ZC61  0.192913  0.412485  0.146354     ...        0.108844    0.197048   \n",
       "\n",
       "        A0A0F7R444  A0A0F7R416  A0A0F7R3Y0  A0A0F7R3S6  A0A0F7R3M8  \\\n",
       "Q9ZFR9    0.130057    0.122946    0.127637    0.132757    0.120328   \n",
       "Q9ZC81    0.079440    0.098909    0.098512    0.070771    0.110741   \n",
       "Q9ZC68    0.112201    0.107268    0.112824    0.115185    0.105061   \n",
       "Q9ZC63    0.081374    0.091124    0.094050    0.078381    0.103397   \n",
       "Q9ZC61    0.084115    0.108381    0.109326    0.065064    0.120763   \n",
       "\n",
       "        A0A0F7R3M4  A0A0F7R3D8  A0A0F7R2X5  \n",
       "Q9ZFR9    0.130324    0.136757    0.129596  \n",
       "Q9ZC81    0.165358    0.193495    0.077961  \n",
       "Q9ZC68    0.129916    0.143231    0.112091  \n",
       "Q9ZC63    0.159618    0.182492    0.078383  \n",
       "Q9ZC61    0.180875    0.205023    0.081547  \n",
       "\n",
       "[5 rows x 2284 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate dissimilarity distances: complement of min-max scaled scores\n",
    "norm_scores = (scores - scores.min(axis=1)) / (scores.max(axis=1) - scores.min(axis=1))\n",
    "distance = 1 - norm_scores\n",
    "\n",
    "distance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare positive pairs and tag species to each protein\n",
    "positive_pairs = df_pos.drop('Pathogen', axis=1).values.tolist()\n",
    "\n",
    "d = df_pos[['Pathogen', 'Pathogen_Uniprot_ID']].values\n",
    "species_tag = {protein: species for species, protein in d}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2284/2284 [01:05<00:00, 35.01it/s]\n"
     ]
    }
   ],
   "source": [
    "# DeNovo negative sampling\n",
    "# T: dissimilarity threshold (only consider proteins with distance < T)\n",
    "T = 0.8\n",
    "\n",
    "negative_pairs = []\n",
    "for pathogen_prot in tqdm(prot_IDs):\n",
    "    # Set up: pathogen protein count and tagging\n",
    "    c = df_pos.Pathogen_Uniprot_ID.tolist().count(pathogen_prot)\n",
    "    tag = species_tag[pathogen_prot]\n",
    "    \n",
    "    # Get a list of pathogen interactors included for sampling\n",
    "    cond = distance.loc[pathogen_prot] > 0.8\n",
    "    patho_inc = cond[cond].index\n",
    "    \n",
    "    # Get list of human proteins that interacts with patho_inc\n",
    "    df = df_pos[df_pos.Pathogen_Uniprot_ID.isin(patho_inc)]\n",
    "    human_inc = set(df.Human_Uniprot_ID)\n",
    "    \n",
    "    # Pair up the current pathogen protein with human_inc\n",
    "    temp_pairs = []\n",
    "    for pair in itertools.product([pathogen_prot], human_inc):\n",
    "        if list(pair) not in positive_pairs:\n",
    "            temp_pairs.append((tag, *pair))\n",
    "    \n",
    "    negative_pairs += temp_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacillus anthracis:\n",
      "113032 non-interacting pairs involving 790 pathogen proteins and 3141 human proteins\n",
      "\n",
      "Francisella tularensis SUBSPECIES TULARENSIS SCHU S4:\n",
      "34512 non-interacting pairs involving 278 pathogen proteins and 284 human proteins\n",
      "\n",
      "Yersinia pestis:\n",
      "151383 non-interacting pairs involving 999 pathogen proteins and 3166 human proteins\n",
      "\n",
      "TOTAL:\n",
      "298927 non-interacting pairs involving 2067 pathogen proteins and 3166 human proteins\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Examine negative pairs DataFrame\n",
    "df_neg = pd.DataFrame(negative_pairs, columns=['Pathogen', 'Pathogen_Uniprot_ID', 'Human_Uniprot_ID'])\n",
    "\n",
    "print_status(df_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathogen</th>\n",
       "      <th>Pathogen_Uniprot_ID</th>\n",
       "      <th>Human_Uniprot_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>Q14686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>Q14571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>P30480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>Q01459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>P32455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pathogen Pathogen_Uniprot_ID Human_Uniprot_ID\n",
       "0  Yersinia pestis              Q9ZFR9           Q14686\n",
       "1  Yersinia pestis              Q9ZFR9           Q14571\n",
       "2  Yersinia pestis              Q9ZFR9           P30480\n",
       "3  Yersinia pestis              Q9ZFR9           Q01459\n",
       "4  Yersinia pestis              Q9ZFR9           P32455"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save negative pairs\n",
    "f_out = os.path.join(dir_out, 'negative_pairs_T%.2f.tsv' % T)\n",
    "df_neg.to_csv(f_out, sep='\\t', index=False)\n",
    "df_neg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr></hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
