{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DeNovo Negative Dataset Sampling for HP-PPI\n",
    "\n",
    "Dissimilarity-based negative sampling for non-interacting protein pairs\n",
    "\n",
    "Reference:\n",
    "\n",
    ">Eid, F. E., Elhefnawi, M., & Heath, L. S. (2016). DeNovo: Virus-host sequence-based protein-protein interaction prediction. *Bioinformatics*, **32**(8), 1144â€“1150. https://doi.org/10.1093/bioinformatics/btv737"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import random\n",
    "\n",
    "from time import time\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "from multiprocessing import Pool\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import PairwiseAligner\n",
    "from Bio.SubsMat.MatrixInfo import blosum30\n",
    "\n",
    "# Print status of datasets\n",
    "def print_status(df):\n",
    "    \n",
    "    # For each pathogen\n",
    "    for pathogen in sorted(set(df.Pathogen)):\n",
    "        df_patho = df[df.Pathogen == pathogen]\n",
    "        i = len(df_patho)\n",
    "        p = len(set(df_patho.Pathogen_Uniprot_ID))\n",
    "        h = len(set(df_patho.Human_Uniprot_ID))\n",
    "        print('%s:\\n%i non-interacting pairs involving %i pathogen proteins and %i human proteins\\n' % (pathogen, i, p, h))\n",
    "    \n",
    "    # Total\n",
    "    i = len(df)\n",
    "    p = len(set(df.Pathogen_Uniprot_ID))\n",
    "    h = len(set(df.Human_Uniprot_ID))\n",
    "    print('TOTAL:\\n%i non-interacting pairs involving %i pathogen proteins and %i human proteins\\n' % (i, p, h))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_dir = os.path.dirname(os.getcwd())\n",
    "dir_in = dir_out = os.path.join(parent_dir, 'data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up the sequences of pathogen proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtained 2284 pathogen proteins for alignment\n"
     ]
    }
   ],
   "source": [
    "# Read positive dataset\n",
    "f_in = os.path.join(dir_in, 'positive_pairs.tsv')\n",
    "df_pos = pd.read_csv(f_in, sep='\\t')\n",
    "\n",
    "# Get pathogen protein IDs for alignment\n",
    "prot_IDs = list(set(df_pos.Pathogen_Uniprot_ID))\n",
    "prot_IDs.sort(reverse=True)\n",
    "\n",
    "print('Obtained %i pathogen proteins for alignment' % len(prot_IDs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtained 2284 pathogen protein sequences\n"
     ]
    }
   ],
   "source": [
    "# Extract pathogen protein sequences\n",
    "f_in = os.path.join(dir_in, 'uniprot_sequences.fasta')\n",
    "seqdict = {} # store sequences in a dict\n",
    "\n",
    "for record in SeqIO.parse(f_in, 'fasta'):\n",
    "    prot_id = record.id.split('|')[1]\n",
    "    \n",
    "    # Check if record is in the pathogen IDs\n",
    "    if prot_id in prot_IDs:\n",
    "        seqdict[prot_id] = record.seq\n",
    "\n",
    "print('Obtained %i pathogen protein sequences' % len(seqdict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All-vs-all global alignment of pathogen proteins\n",
    "\n",
    "May skip to the next section (`Negative Sampling`) if alignment scores file already exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Needleman-Wunsch'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set up aligner\n",
    "aligner = PairwiseAligner()\n",
    "\n",
    "# Gap scoring: linear, 8 (as in MATLAB, which is the platform used by Eid et al., 2016)\n",
    "# Matrix: BLOSUM30 to capture distant relationships\n",
    "\n",
    "aligner.gap_score = -8\n",
    "aligner.substitution_matrix = blosum30\n",
    "aligner.algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated 2609470 pairs for alignment\n"
     ]
    }
   ],
   "source": [
    "# Generate pairs of proteins to be aligned\n",
    "pairs = [pair for pair in itertools.combinations_with_replacement(prot_IDs, 2)]\n",
    "print('Generated %i pairs for alignment' % len(pairs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2284, 2284)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <th>Q9ZC81</th>\n",
       "      <th>Q9ZC68</th>\n",
       "      <th>Q9ZC63</th>\n",
       "      <th>Q9ZC61</th>\n",
       "      <th>Q9ZC54</th>\n",
       "      <th>Q9ZC51</th>\n",
       "      <th>Q9ZC50</th>\n",
       "      <th>Q9ZC30</th>\n",
       "      <th>Q9ZC29</th>\n",
       "      <th>...</th>\n",
       "      <th>A0A0F7R517</th>\n",
       "      <th>A0A0F7R4Q5</th>\n",
       "      <th>A0A0F7R444</th>\n",
       "      <th>A0A0F7R416</th>\n",
       "      <th>A0A0F7R3Y0</th>\n",
       "      <th>A0A0F7R3S6</th>\n",
       "      <th>A0A0F7R3M8</th>\n",
       "      <th>A0A0F7R3M4</th>\n",
       "      <th>A0A0F7R3D8</th>\n",
       "      <th>A0A0F7R2X5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC81</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC68</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC63</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC61</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Q9ZFR9  Q9ZC81  Q9ZC68  Q9ZC63  Q9ZC61  Q9ZC54  Q9ZC51  Q9ZC50  \\\n",
       "Q9ZFR9     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC81     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC68     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC63     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Q9ZC61     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "        Q9ZC30  Q9ZC29     ...      A0A0F7R517  A0A0F7R4Q5  A0A0F7R444  \\\n",
       "Q9ZFR9     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC81     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC68     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC63     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "Q9ZC61     0.0     0.0     ...             0.0         0.0         0.0   \n",
       "\n",
       "        A0A0F7R416  A0A0F7R3Y0  A0A0F7R3S6  A0A0F7R3M8  A0A0F7R3M4  \\\n",
       "Q9ZFR9         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC81         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC68         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC63         0.0         0.0         0.0         0.0         0.0   \n",
       "Q9ZC61         0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "        A0A0F7R3D8  A0A0F7R2X5  \n",
       "Q9ZFR9         0.0         0.0  \n",
       "Q9ZC81         0.0         0.0  \n",
       "Q9ZC68         0.0         0.0  \n",
       "Q9ZC63         0.0         0.0  \n",
       "Q9ZC61         0.0         0.0  \n",
       "\n",
       "[5 rows x 2284 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize zeros DataFrame for storing alignment scores\n",
    "scores = pd.DataFrame(0., index=prot_IDs, columns=prot_IDs)\n",
    "\n",
    "scores.shape\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get the alignment score\n",
    "def get_alignment_score(pair):\n",
    "    score = aligner.score(*map(seqdict.get, pair))\n",
    "    \n",
    "    # Return both proteins along with the alignment score\n",
    "    return (*pair, score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment duration: 11.8951 minutes\n"
     ]
    }
   ],
   "source": [
    "# Perform alignments with parallelization\n",
    "with Pool(4) as p:\n",
    "    t0 = time()\n",
    "    \n",
    "    results = p.map(get_alignment_score, pairs)\n",
    "    \n",
    "    t = time() - t0\n",
    "    print('Alignment duration: %.4f minutes' % (t/60))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2609470/2609470 [00:57<00:00, 45378.20it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <th>Q9ZC81</th>\n",
       "      <th>Q9ZC68</th>\n",
       "      <th>Q9ZC63</th>\n",
       "      <th>Q9ZC61</th>\n",
       "      <th>Q9ZC54</th>\n",
       "      <th>Q9ZC51</th>\n",
       "      <th>Q9ZC50</th>\n",
       "      <th>Q9ZC30</th>\n",
       "      <th>Q9ZC29</th>\n",
       "      <th>...</th>\n",
       "      <th>A0A0F7R517</th>\n",
       "      <th>A0A0F7R4Q5</th>\n",
       "      <th>A0A0F7R444</th>\n",
       "      <th>A0A0F7R416</th>\n",
       "      <th>A0A0F7R3Y0</th>\n",
       "      <th>A0A0F7R3S6</th>\n",
       "      <th>A0A0F7R3M8</th>\n",
       "      <th>A0A0F7R3M4</th>\n",
       "      <th>A0A0F7R3D8</th>\n",
       "      <th>A0A0F7R2X5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <td>3953.0</td>\n",
       "      <td>-964.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>-793.0</td>\n",
       "      <td>-1272.0</td>\n",
       "      <td>-374.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>-964.0</td>\n",
       "      <td>-198.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-685.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>-975.0</td>\n",
       "      <td>-552.0</td>\n",
       "      <td>-561.0</td>\n",
       "      <td>-1411.0</td>\n",
       "      <td>-402.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>-999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC81</th>\n",
       "      <td>-964.0</td>\n",
       "      <td>2368.0</td>\n",
       "      <td>-580.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>-169.0</td>\n",
       "      <td>-1394.0</td>\n",
       "      <td>-889.0</td>\n",
       "      <td>-2941.0</td>\n",
       "      <td>-275.0</td>\n",
       "      <td>...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>-1002.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>-90.0</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>-73.0</td>\n",
       "      <td>-229.0</td>\n",
       "      <td>-613.0</td>\n",
       "      <td>-985.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC68</th>\n",
       "      <td>25.0</td>\n",
       "      <td>-580.0</td>\n",
       "      <td>3479.0</td>\n",
       "      <td>-423.0</td>\n",
       "      <td>-867.0</td>\n",
       "      <td>-51.0</td>\n",
       "      <td>-236.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>-1444.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-278.0</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>-601.0</td>\n",
       "      <td>-245.0</td>\n",
       "      <td>-278.0</td>\n",
       "      <td>-1036.0</td>\n",
       "      <td>-116.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>-631.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC63</th>\n",
       "      <td>-793.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-423.0</td>\n",
       "      <td>2311.0</td>\n",
       "      <td>-187.0</td>\n",
       "      <td>-50.0</td>\n",
       "      <td>-1243.0</td>\n",
       "      <td>-618.0</td>\n",
       "      <td>-2708.0</td>\n",
       "      <td>-129.0</td>\n",
       "      <td>...</td>\n",
       "      <td>80.0</td>\n",
       "      <td>-816.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>-251.0</td>\n",
       "      <td>-82.0</td>\n",
       "      <td>-514.0</td>\n",
       "      <td>-747.0</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC61</th>\n",
       "      <td>-1272.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>-867.0</td>\n",
       "      <td>-187.0</td>\n",
       "      <td>2246.0</td>\n",
       "      <td>-440.0</td>\n",
       "      <td>-1772.0</td>\n",
       "      <td>-1185.0</td>\n",
       "      <td>-3294.0</td>\n",
       "      <td>-540.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-125.0</td>\n",
       "      <td>-1313.0</td>\n",
       "      <td>-59.0</td>\n",
       "      <td>-280.0</td>\n",
       "      <td>-234.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-428.0</td>\n",
       "      <td>-954.0</td>\n",
       "      <td>-1215.0</td>\n",
       "      <td>-38.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Q9ZFR9  Q9ZC81  Q9ZC68  Q9ZC63  Q9ZC61  Q9ZC54  Q9ZC51  Q9ZC50  \\\n",
       "Q9ZFR9  3953.0  -964.0    25.0  -793.0 -1272.0  -374.0     4.0   122.0   \n",
       "Q9ZC81  -964.0  2368.0  -580.0     2.0    41.0  -169.0 -1394.0  -889.0   \n",
       "Q9ZC68    25.0  -580.0  3479.0  -423.0  -867.0   -51.0  -236.0    87.0   \n",
       "Q9ZC63  -793.0     2.0  -423.0  2311.0  -187.0   -50.0 -1243.0  -618.0   \n",
       "Q9ZC61 -1272.0    41.0  -867.0  -187.0  2246.0  -440.0 -1772.0 -1185.0   \n",
       "\n",
       "        Q9ZC30  Q9ZC29     ...      A0A0F7R517  A0A0F7R4Q5  A0A0F7R444  \\\n",
       "Q9ZFR9  -964.0  -198.0     ...          -685.0       115.0      -975.0   \n",
       "Q9ZC81 -2941.0  -275.0     ...            37.0     -1002.0        49.0   \n",
       "Q9ZC68 -1444.0    29.0     ...          -278.0       -13.0      -601.0   \n",
       "Q9ZC63 -2708.0  -129.0     ...            80.0      -816.0        12.0   \n",
       "Q9ZC61 -3294.0  -540.0     ...          -125.0     -1313.0       -59.0   \n",
       "\n",
       "        A0A0F7R416  A0A0F7R3Y0  A0A0F7R3S6  A0A0F7R3M8  A0A0F7R3M4  \\\n",
       "Q9ZFR9      -552.0      -561.0     -1411.0      -402.0       104.0   \n",
       "Q9ZC81       -90.0        -7.0       -73.0      -229.0      -613.0   \n",
       "Q9ZC68      -245.0      -278.0     -1036.0      -116.0        63.0   \n",
       "Q9ZC63        75.0        91.0      -251.0       -82.0      -514.0   \n",
       "Q9ZC61      -280.0      -234.0        19.0      -428.0      -954.0   \n",
       "\n",
       "        A0A0F7R3D8  A0A0F7R2X5  \n",
       "Q9ZFR9       146.0      -999.0  \n",
       "Q9ZC81      -985.0        36.0  \n",
       "Q9ZC68        35.0      -631.0  \n",
       "Q9ZC63      -747.0        73.0  \n",
       "Q9ZC61     -1215.0       -38.0  \n",
       "\n",
       "[5 rows x 2284 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store alignment scores\n",
    "for prot1, prot2, score in tqdm(results):\n",
    "    scores.at[prot1, prot2] = score\n",
    "    scores.at[prot2, prot1] = score\n",
    "\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the scores DataFrame\n",
    "f_out = os.path.join(dir_out, 'blosum30_global_scores.tsv')\n",
    "scores.to_csv(f_out, sep='\\t', index=prot_IDs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Negative sampling\n",
    "\n",
    "Select non-interacting pairs based on dissimilarity between pathogen protein interactors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the scores DataFrame\n",
    "f_in = os.path.join(dir_in, 'blosum30_global_scores.tsv')\n",
    "scores = pd.read_csv(f_in, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removed outlier protein: Q81SN0\n",
      "Removed outlier protein: Q8CZU2\n",
      "Removed outlier protein: Q7CGD9\n",
      "Removed outlier protein: Q9Z373\n",
      "Removed outlier protein: Q7CGR6\n",
      "Removed outlier protein: Q7CFY4\n",
      "Removed outlier protein: Q8D0R8\n",
      "No more outliers found.\n"
     ]
    }
   ],
   "source": [
    "# Designate outlier scores as NaN (protein with very low alignment score compared to others)\n",
    "# Keep removing outliers as long as it has minimum scores across x% of total pathogen proteins\n",
    "x = 0.8\n",
    "\n",
    "while True:\n",
    "    row_mins = scores.idxmin(axis=1)\n",
    "    outlier_id = row_mins.mode()[0] # a bug: axis=0 actually considers columns instead of rows\n",
    "    \n",
    "    n_outlier = Counter(row_mins)[outlier_id]\n",
    "    if n_outlier > int(x * len(prot_IDs)):\n",
    "        \n",
    "        # Remove all entries in the outlier column except for its own row\n",
    "        scores.loc[~(scores.index == outlier_id), outlier_id] = np.nan\n",
    "        print('Removed outlier protein: %s' % outlier_id)\n",
    "    \n",
    "    else:\n",
    "        print('No more outliers found.')\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <th>Q9ZC81</th>\n",
       "      <th>Q9ZC68</th>\n",
       "      <th>Q9ZC63</th>\n",
       "      <th>Q9ZC61</th>\n",
       "      <th>Q9ZC54</th>\n",
       "      <th>Q9ZC51</th>\n",
       "      <th>Q9ZC50</th>\n",
       "      <th>Q9ZC30</th>\n",
       "      <th>Q9ZC29</th>\n",
       "      <th>...</th>\n",
       "      <th>A0A0F7R517</th>\n",
       "      <th>A0A0F7R4Q5</th>\n",
       "      <th>A0A0F7R444</th>\n",
       "      <th>A0A0F7R416</th>\n",
       "      <th>A0A0F7R3Y0</th>\n",
       "      <th>A0A0F7R3S6</th>\n",
       "      <th>A0A0F7R3M8</th>\n",
       "      <th>A0A0F7R3M4</th>\n",
       "      <th>A0A0F7R3D8</th>\n",
       "      <th>A0A0F7R2X5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9ZFR9</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.198653</td>\n",
       "      <td>0.217028</td>\n",
       "      <td>0.188430</td>\n",
       "      <td>0.206068</td>\n",
       "      <td>0.199170</td>\n",
       "      <td>0.277767</td>\n",
       "      <td>0.215522</td>\n",
       "      <td>0.466832</td>\n",
       "      <td>0.209832</td>\n",
       "      <td>...</td>\n",
       "      <td>0.208260</td>\n",
       "      <td>0.213014</td>\n",
       "      <td>0.192820</td>\n",
       "      <td>0.189946</td>\n",
       "      <td>0.196090</td>\n",
       "      <td>0.191968</td>\n",
       "      <td>0.189010</td>\n",
       "      <td>0.213020</td>\n",
       "      <td>0.229633</td>\n",
       "      <td>0.192252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC81</th>\n",
       "      <td>0.314648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.255042</td>\n",
       "      <td>0.140169</td>\n",
       "      <td>0.129159</td>\n",
       "      <td>0.186470</td>\n",
       "      <td>0.369265</td>\n",
       "      <td>0.280584</td>\n",
       "      <td>0.603017</td>\n",
       "      <td>0.214576</td>\n",
       "      <td>...</td>\n",
       "      <td>0.164727</td>\n",
       "      <td>0.285335</td>\n",
       "      <td>0.131653</td>\n",
       "      <td>0.161519</td>\n",
       "      <td>0.162248</td>\n",
       "      <td>0.113179</td>\n",
       "      <td>0.178268</td>\n",
       "      <td>0.258075</td>\n",
       "      <td>0.301957</td>\n",
       "      <td>0.130284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC68</th>\n",
       "      <td>0.251360</td>\n",
       "      <td>0.175759</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.165969</td>\n",
       "      <td>0.182345</td>\n",
       "      <td>0.179160</td>\n",
       "      <td>0.293475</td>\n",
       "      <td>0.217775</td>\n",
       "      <td>0.499897</td>\n",
       "      <td>0.195848</td>\n",
       "      <td>...</td>\n",
       "      <td>0.183720</td>\n",
       "      <td>0.221301</td>\n",
       "      <td>0.170480</td>\n",
       "      <td>0.171056</td>\n",
       "      <td>0.178803</td>\n",
       "      <td>0.169886</td>\n",
       "      <td>0.171251</td>\n",
       "      <td>0.215596</td>\n",
       "      <td>0.236731</td>\n",
       "      <td>0.170219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC63</th>\n",
       "      <td>0.303705</td>\n",
       "      <td>0.141060</td>\n",
       "      <td>0.245178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142514</td>\n",
       "      <td>0.179098</td>\n",
       "      <td>0.359382</td>\n",
       "      <td>0.263144</td>\n",
       "      <td>0.586967</td>\n",
       "      <td>0.205582</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162134</td>\n",
       "      <td>0.273292</td>\n",
       "      <td>0.133863</td>\n",
       "      <td>0.151366</td>\n",
       "      <td>0.156261</td>\n",
       "      <td>0.123660</td>\n",
       "      <td>0.169140</td>\n",
       "      <td>0.251854</td>\n",
       "      <td>0.286737</td>\n",
       "      <td>0.128068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9ZC61</th>\n",
       "      <td>0.334357</td>\n",
       "      <td>0.138735</td>\n",
       "      <td>0.273076</td>\n",
       "      <td>0.151642</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.203259</td>\n",
       "      <td>0.394005</td>\n",
       "      <td>0.299633</td>\n",
       "      <td>0.627333</td>\n",
       "      <td>0.230902</td>\n",
       "      <td>...</td>\n",
       "      <td>0.174495</td>\n",
       "      <td>0.305471</td>\n",
       "      <td>0.138104</td>\n",
       "      <td>0.173209</td>\n",
       "      <td>0.176115</td>\n",
       "      <td>0.107761</td>\n",
       "      <td>0.190624</td>\n",
       "      <td>0.279502</td>\n",
       "      <td>0.316665</td>\n",
       "      <td>0.134714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Q9ZFR9    Q9ZC81    Q9ZC68    Q9ZC63    Q9ZC61    Q9ZC54    Q9ZC51  \\\n",
       "Q9ZFR9  0.000000  0.198653  0.217028  0.188430  0.206068  0.199170  0.277767   \n",
       "Q9ZC81  0.314648  0.000000  0.255042  0.140169  0.129159  0.186470  0.369265   \n",
       "Q9ZC68  0.251360  0.175759  0.000000  0.165969  0.182345  0.179160  0.293475   \n",
       "Q9ZC63  0.303705  0.141060  0.245178  0.000000  0.142514  0.179098  0.359382   \n",
       "Q9ZC61  0.334357  0.138735  0.273076  0.151642  0.000000  0.203259  0.394005   \n",
       "\n",
       "          Q9ZC50    Q9ZC30    Q9ZC29     ...      A0A0F7R517  A0A0F7R4Q5  \\\n",
       "Q9ZFR9  0.215522  0.466832  0.209832     ...        0.208260    0.213014   \n",
       "Q9ZC81  0.280584  0.603017  0.214576     ...        0.164727    0.285335   \n",
       "Q9ZC68  0.217775  0.499897  0.195848     ...        0.183720    0.221301   \n",
       "Q9ZC63  0.263144  0.586967  0.205582     ...        0.162134    0.273292   \n",
       "Q9ZC61  0.299633  0.627333  0.230902     ...        0.174495    0.305471   \n",
       "\n",
       "        A0A0F7R444  A0A0F7R416  A0A0F7R3Y0  A0A0F7R3S6  A0A0F7R3M8  \\\n",
       "Q9ZFR9    0.192820    0.189946    0.196090    0.191968    0.189010   \n",
       "Q9ZC81    0.131653    0.161519    0.162248    0.113179    0.178268   \n",
       "Q9ZC68    0.170480    0.171056    0.178803    0.169886    0.171251   \n",
       "Q9ZC63    0.133863    0.151366    0.156261    0.123660    0.169140   \n",
       "Q9ZC61    0.138104    0.173209    0.176115    0.107761    0.190624   \n",
       "\n",
       "        A0A0F7R3M4  A0A0F7R3D8  A0A0F7R2X5  \n",
       "Q9ZFR9    0.213020    0.229633    0.192252  \n",
       "Q9ZC81    0.258075    0.301957    0.130284  \n",
       "Q9ZC68    0.215596    0.236731    0.170219  \n",
       "Q9ZC63    0.251854    0.286737    0.128068  \n",
       "Q9ZC61    0.279502    0.316665    0.134714  \n",
       "\n",
       "[5 rows x 2284 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate dissimilarity distances: complement of min-max scaled scores\n",
    "norm_scores = (scores - scores.min(axis=1)) / (scores.max(axis=1) - scores.min(axis=1))\n",
    "distance = 1 - norm_scores\n",
    "\n",
    "distance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare positive pairs and tag species to each protein\n",
    "positive_pairs = df_pos.drop('Pathogen', axis=1).values.tolist()\n",
    "\n",
    "d = df_pos[['Pathogen', 'Pathogen_Uniprot_ID']].values\n",
    "species_tag = {protein: species for species, protein in d}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2284/2284 [01:59<00:00, 19.08it/s]\n"
     ]
    }
   ],
   "source": [
    "# DeNovo negative sampling\n",
    "# T: dissimilarity threshold (only consider proteins with distance < T)\n",
    "# n: relative size of negative pairs for a particular protein to be generated compared to its positive interactions\n",
    "T = 0.8\n",
    "n = 5\n",
    "\n",
    "negative_pairs = []\n",
    "for pathogen_prot in tqdm(prot_IDs):\n",
    "    # Set up: pathogen protein count and tagging\n",
    "    c = df_pos.Pathogen_Uniprot_ID.tolist().count(pathogen_prot)\n",
    "    tag = species_tag[pathogen_prot]\n",
    "    \n",
    "    # Get a list of pathogen interactors included for sampling\n",
    "    cond = distance.loc[pathogen_prot] > 0.8\n",
    "    patho_inc = cond[cond].index\n",
    "    \n",
    "    # Get list of human proteins that interacts with patho_inc\n",
    "    df = df_pos[df_pos.Pathogen_Uniprot_ID.isin(patho_inc)]\n",
    "    human_inc = set(df.Human_Uniprot_ID)\n",
    "    \n",
    "    # Pair up the current pathogen protein with human_inc\n",
    "    temp_pairs = []\n",
    "    for pair in itertools.product([pathogen_prot], human_inc):\n",
    "        if list(pair) not in positive_pairs:\n",
    "            temp_pairs.append((tag, *pair))\n",
    "    \n",
    "    # Generate N negative pairs\n",
    "    N = n * c\n",
    "    \n",
    "    if len(temp_pairs) > N:\n",
    "        negative_pairs += random.sample(temp_pairs, N)\n",
    "    else:\n",
    "        negative_pairs += temp_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacillus anthracis:\n",
      "10777 non-interacting pairs involving 718 pathogen proteins and 1120 human proteins\n",
      "\n",
      "Francisella tularensis SUBSPECIES TULARENSIS SCHU S4:\n",
      "3500 non-interacting pairs involving 243 pathogen proteins and 465 human proteins\n",
      "\n",
      "Yersinia pestis:\n",
      "13137 non-interacting pairs involving 880 pathogen proteins and 789 human proteins\n",
      "\n",
      "TOTAL:\n",
      "27414 non-interacting pairs involving 1841 pathogen proteins and 1307 human proteins\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Examine negative pairs DataFrame\n",
    "df_neg = pd.DataFrame(negative_pairs, columns=['Pathogen',\n",
    "                                               'Pathogen_Uniprot_ID',\n",
    "                                               'Human_Uniprot_ID'])\n",
    "\n",
    "print_status(df_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathogen</th>\n",
       "      <th>Pathogen_Uniprot_ID</th>\n",
       "      <th>Human_Uniprot_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>P01857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>Q9ULF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>O95236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>P46379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Q9ZFR9</td>\n",
       "      <td>Q14192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pathogen Pathogen_Uniprot_ID Human_Uniprot_ID\n",
       "0  Yersinia pestis              Q9ZFR9           P01857\n",
       "1  Yersinia pestis              Q9ZFR9           Q9ULF5\n",
       "2  Yersinia pestis              Q9ZFR9           O95236\n",
       "3  Yersinia pestis              Q9ZFR9           P46379\n",
       "4  Yersinia pestis              Q9ZFR9           Q14192"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save negative pairs\n",
    "f_out = os.path.join(dir_out, 'negative_pairs.tsv')\n",
    "df_neg.to_csv(f_out, sep='\\t', index=False)\n",
    "df_neg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr></hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
